solidity/contracts/converter/types/fixed-rate-pool/FixedRatePoolConverter.sol

```diff
@@ -0,0 +1,315 @@
+     *
+     * @return expected amount in units of the target reserve token
+     * @return expected fee in units of the target reserve token
+     */
+    function targetAmountAndFee(
+        IERC20Token _sourceToken,
+        IERC20Token _targetToken,
+        uint256 _amount
+    ) public view override active returns (uint256, uint256) {
+        uint128[2] memory factors = [_rateN, _rateD];
+
+        uint256 sourceId = __reserveIds[_sourceToken] - 1;
+        uint256 targetId = __reserveIds[_targetToken] - 1;
+    
+        uint256 amount = _amount.mul(factors[sourceId]).div(factors[targetId]);
+
+        uint256 fee = calculateFee(amount);
+
+        return (amount - fee, fee);
+    }
+
+    /**
+     * @dev returns the required amount and expected fee for converting one reserve to another
+     *
+     * @param _sourceToken address of the source reserve token contract
+     * @param _targetToken address of the target reserve token contract
+     * @param _amount      amount of target reserve tokens desired
+     *
+     * @return required amount in units of the source reserve token
+     * @return expected fee in units of the target reserve token
+     */
+    function sourceAmountAndFee(
+        IERC20Token _sourceToken,
+        IERC20Token _targetToken,
+        uint256 _amount
+    ) public view override active returns (uint256, uint256) {
+        uint128[2] memory factors = [_rateN, _rateD];
+
+        uint256 sourceId = __reserveIds[_sourceToken] - 1;
+        uint256 targetId = __reserveIds[_targetToken] - 1;
+    
+        uint256 fee = calculateFeeInv(_amount);
+
+        uint256 amount = _amount.add(fee).mul(factors[targetId]).div(factors[sourceId]);
+
+        return (amount, fee);
+    }
+
+    /**
+     * @dev converts a specific amount of source tokens to target tokens
+     *
+     * @param _sourceToken source ERC20 token
+     * @param _targetToken target ERC20 token
+     * @param _amount      amount of tokens to convert (in units of the source token)
+     * @param _trader      address of the caller who executed the conversion
+     * @param _beneficiary wallet to receive the conversion result
+     *
+     * @return amount of tokens received (in units of the target token)
+     */
+    function doConvert(
+        IERC20Token _sourceToken,
+        IERC20Token _targetToken,
+        uint256 _amount,
+        address _trader,
+        address payable _beneficiary
+    ) internal override returns (uint256) {
+        // update the recent average rate
+        updateRecentAverageRate();
+
+        uint256 sourceId = __reserveIds[_sourceToken];
+        uint256 targetId = __reserveIds[_targetToken];
+
+        (uint256 sourceBalance, uint256 targetBalance) = reserveBalances(sourceId, targetId);
+        uint256 targetAmount;
+        {
+            uint128[2] memory factors = [_rateN, _rateD];
+            targetAmount = _amount.mul(factors[sourceId - 1]).div(factors[targetId - 1]);
```

Comment:
why not use `targetAmountAndFee` here?

Reply:
Fixed.

Reply:
Any reason to even override that function and not just use the parent's implementation?

Reply:
Yes, the parent implementation is different.

Reply:
I don't see the `doConvert` override anymore - gone?

Reply:
Yes, I've replaced the two large functions (doConvert and addLiquidity) with a few smaller ones.

Reply:
nice!